### ISSUE TITLE
Sync is very, very slow.

### ISSUE THREAD

**Issue by eperdeme** [containers/skopeo#1021]
Hello,

Does anybody have any thoughts on how to speed up a sync. We mirror a lot of external repos on mass for all versions and the process can take several hours layer by layer, tag by tag, repo by repo.

Is there anyway to parallel to run for every repo ? I've thought about maybe splitting each external repo in to separate yamls and run more skopeo commands at same time.

```
skopeo --insecure-policy sync --src yaml --dest docker src_mirrors.yaml dest.internal.repo/mirrors/ --scoped --dest-creds X:X

docker.io:
    images:
        istio/proxyv2: []
        sergrua/kube-tagger: []
        splunk/fluentd-hec: []
        hazelcast/management-center: []
        k8srestdev/scaling: []
        hjacobs/kube-resource-report: []
        tutum/dnsutils: []
        bitnami/external-dns: []
        falcosecurity/falco: []
        praqma/helmsman: []
        weaveworks/flagger: []
        weaveworks/flagger-loadtester: []
quay.io:
    images:
        coreos/prometheus-config-reloader: []
        prometheus/prometheus: []
        prometheus/alertmanager: []
        prometheus/node-exporter: []
        kiali/kiali: []
        coreos/configmap-reload: []
        coreos/prometheus-operator: []
        thanos/thanos: []
        calico/node: []
        coreos/kube-state-metrics: []
        calico/typha: []
        external_storage/efs-provisioner: []
gcr.io:
    images: 
        kubernetes-helm/tiller: []
        google_containers/busybox: []
        k8s-artifacts-prod/autoscaling: []
k8s.gcr.io:
    images:
        cluster-proportional-autoscaler-amd64: []
        metrics-server-amd64: [] 
```

**Comment by mtrmac** [comment]
Thanks for your report.

You might try splitting the YAML and running that in parallel outside of `skopeo sync`, at least to get an initial guess at the benefit, if any.

>  the process can take several hours layer by layer, tag by tag, repo by repo.

I mean, if there is _that much_ data to copy, there’s _that much_ data to copy, and ultimately I’d expect the sync to be link-speed bound.

Or are you observing that the same images are being copied all over again, even when running the same `skopeo sync` session back to back with no changes at the source? That might be a bug — but please make sure that the blob info cache ( https://github.com/containers/image/blob/22129f19a77890293540e67843f28860f5a77895/pkg/blobinfocache/default.go#L25 ) is preserved across runs of `skopeo sync`, if you are maybe running it from a container that is torn down after each sync, or something like that.

**Comment by eperdeme** [comment]
You're right about the cache bit seemingly not working. If I run it from my mac  back to back it seems to try and resend every layer.

```
➜  docker-mirror git:(master) ✗ skopeo --insecure-policy sync --src yaml --dest docker src_mirrors.yaml internal.mirror.local/mirrors/ --scoped --dest-creds svcSqa.trd.tpt:November17#

INFO[0000] Processing repo                               registry=quay.io repo=kiali/kiali
INFO[0000] Querying registry for image tags              registry=quay.io repo=kiali/kiali
INFO[0000] Getting tags                                  image=quay.io/kiali/kiali
INFO[0002] Copying image tag 1/130                       from="docker://quay.io/kiali/kiali:v0.18.0" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.18.0"
Getting image source signatures
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 7dc0dca2b151 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 902d2f3f25bc skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 0ef2a084a469 skipped: already exists  
Copying blob a444b6991f77 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 55ce4b02a7cb .                                        
Writing manifest to image destination
Storing signatures
INFO[0004] Copying image tag 2/130                       from="docker://quay.io/kiali/kiali:v0.19.0-snapshot.0" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.19.0-snapshot.0"
Getting image source signatures
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 8ba884070f61 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 07fc895d6ba6 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob f0c55eb852ce skipped: already exists  
Copying blob 31f1c61558e6 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob e6809ae6b48b .                                        
Writing manifest to image destination
Storing signatures
INFO[0006] Copying image tag 3/130                       from="docker://quay.io/kiali/kiali:v0.19.0-snapshot.1" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.19.0-snapshot.1"
^C
➜  docker-mirror git:(master) ✗ skopeo --insecure-policy sync --src yaml --dest docker src_mirrors.yaml internal.mirror.local/mirrors/ --scoped --dest-creds svcSqa.trd.tpt:November17#

INFO[0000] Processing repo                               registry=quay.io repo=kiali/kiali
INFO[0000] Querying registry for image tags              registry=quay.io repo=kiali/kiali
INFO[0000] Getting tags                                  image=quay.io/kiali/kiali
INFO[0002] Copying image tag 1/130                       from="docker://quay.io/kiali/kiali:v0.18.0" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.18.0"
Getting image source signatures
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 7dc0dca2b151 skipped: already exists  
Copying blob 902d2f3f25bc skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 0ef2a084a469 skipped: already exists  
Copying blob 55ce4b02a7cb skipped: already exists  
Copying blob a444b6991f77 skipped: already exists  
Copying blob a3ed95caeb02 .                                        
Writing manifest to image destination
Storing signatures
INFO[0004] Copying image tag 2/130                       from="docker://quay.io/kiali/kiali:v0.19.0-snapshot.0" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.19.0-snapshot.0"
Getting image source signatures
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob 07fc895d6ba6 skipped: already exists  
Copying blob 8ba884070f61 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob a3ed95caeb02 skipped: already exists  
Copying blob e6809ae6b48b skipped: already exists  
Copying blob f0c55eb852ce skipped: already exists  
Copying blob 31f1c61558e6 skipped: already exists  
Copying blob a3ed95caeb02 .                                        
Writing manifest to image destination
Storing signatures
INFO[0006] Copying image tag 3/130                       from="docker://quay.io/kiali/kiali:v0.19.0-snapshot.1" to="docker://internal.mirror.local/mirrors/quay.io/kiali/kiali:v0.19.0-snapshot.1"

**Comment by mtrmac** [comment]
No, that seems to work well enough; “already exists” means not copying the hundreds of megabytes of data. (Actually that does not even need the cache, the cache could help when a new repository appears.)

It is possible that this is just too many roundtrips, but at 130 images even at 10 layers, even if it took one second per layer, that would hardly add up to “several hours”. Could you be hitting a rate limit at one of the both of the involved registries?

**Comment by eperdeme** [comment]
I think it's just too many round trips. Some images have 4,000 tags so maybe it's just as everything is serial it takes a very long time.
I've split the mirrors in to separate files per src and added regex to avoid pulling down RC/dev tags where possible, so providing we never need to test RCs internally we'll make do this this speed improvement.

**Comment by gjenkins8** [comment]
I have the same sort of problem with sync (which is very nice by the way!). Where the use case is to ensure 'remote' registries are kept in sync with an internal/source registry. We run a period job which 'duplicates' out internal image/tags, but where generally the majority of the content is unmodified.

The below timestamps are for a specific image copy (where the destination image already exists), and which took just over 1s. This is 11 minutes of so (for this set of tags, there are a few more images).
 
From looking at [the source](), it seems `sync` implements a 'copy' for each tag (reasonable of course). But in case there are any possible optimization where the destination is expected to exist unmodified? In particular the writing of the manifest seems to always take ~300ms, but should be unchanged?

```
[2020-08-25T15:45:20.030Z] time="2020-08-25T15:45:19Z" level=info msg="Copying image tag 2/685" from="docker://registry.mycompany.com/myimage:1.104.3700-224" to="docker://123456.dkr.ecr.us-east-1.amazonaws.com/mydept/myimage1.104.3700-224"
[2020-08-25T15:45:20.030Z] Getting image source signatures
[2020-08-25T15:45:20.290Z] Copying blob sha256:1d2c4ce43b78cb9a97ede7f19ad1406a43ee50532568bda660193e4a404b424f
[2020-08-25T15:45:20.290Z] Copying blob sha256:ab36d12d348c0bea37250f280d104b217e4a41eb0138810bf08a4bc2217ebc9a
[2020-08-25T15:45:20.290Z] Copying blob sha256:1bb3415aff5dfdb4d01df8cd5fc5e8e83a020fd94ba26666dfdf8f2210cf1128
[2020-08-25T15:45:20.551Z] Copying blob sha256:8675d36fe29c085721a38ecb2904f1eee2d961f10bbd571da44f285237a9bdc4
[2020-08-25T15:45:20.551Z] Copying blob sha256:36b5a23020d4b97a70093feaa7d20908a9df33a0c25e18b2d25475a602014f4d
[2020-08-25T15:45:20.551Z] Copying blob sha256:1c9f515fc6ab2b7ebfcaffd8af681b68869d78a3b19c69e87c296363ab1bc2fe
[2020-08-25T15:45:20.551Z] Copying blob sha256:8f0a11903a588d724fc210eee6332b02c21d419d44928be199f8221597019b88
[2020-08-25T15:45:20.551Z] Copying blob sha256:e0190a56de136e88d9d03def88bac4be75183c25092c414b7a763baaf69b1366
[2020-08-25T15:45:20.551Z] Copying blob sha256:2da361b4af751864acdb25852798a7111747ba7d8553efe81db13c3c65594e54
[2020-08-25T15:45:20.551Z] Copying blob sha256:022f7122d0d11a32f6494b8b56d43139f79ea412c9fb83ac2bce7a67dd904748
[2020-08-25T15:45:20.551Z] Copying blob sha256:5fb3ed35e1ccebc8b1c114d7dd1ddb5eab87ea8cac4151cf26c6871e2eed102f
[2020-08-25T15:45:20.813Z] Copying blob sha256:1ca6ec59c1f4bad31d655fd2dcc7542d2419170cbb98b18a3965f668cacdf8e2
[2020-08-25T15:45:20.813Z] Copying blob sha256:8686faffd55b78f473033906abbe5304d5484c04c2e129124df32483d1052678
[2020-08-25T15:45:20.813Z] Copying config sha256:eb21d52c12296d28abaae33ee35de840c7d11d0e190f1e45091f5a3325327d14
[2020-08-25T15:45:21.076Z] Writing manifest to image destination
[2020-08-25T15:45:21.076Z] Storing signatures
[2020-08-25T15:45:21.076Z] time="2020-08-25T15:45:21Z" level=info msg="Copying image tag 3/685" from="...

**Comment by mtrmac** [comment]
> I think it's just too many round trips. Some images have 4,000 tags so maybe it's just as everything is serial it takes a very long time.

Ouch. Yeah, in the best possible case reading 4000 manifests to see whether the tags have moved (and that can only be done one tag at a time) is going to take some time. OTOH if most of the time is spent waiting, parallelizing this across a few threads might noticeably help. The code structure right now makes that a bit difficult (the checks are part of a copy, and a copy automatically parallelizes copies of individual layers, so a dozen copies * 6 layer threads = 70 parallel TCP streams), it might make sense to do a “presence” check before starting a copy if _almost all_ images are not added/updated.

**Comment by mtrmac** [comment]
> From looking at [the source](https://github.com/containers/skopeo/blob/master/cmd/skopeo/sync.go#L575), it seems `sync` implements a 'copy' for each tag (reasonable of course). But in case there are any possible optimization where the destination is expected to exist unmodified? In particular the writing of the manifest seems to always take ~300ms, but should be unchanged?

The “Copying blob/Writing manifest” entries record _starts_ of that activity, and we do up to 6 blob copies in parallel. So, going by this specific record, if I’m reading it right, reading the original manifest and config took 0 ms, ??? set up took 260 ms, checks that the layers are present took 523 ms, writing the config 263 ms, writing the manifest 0 ms. Several of those times just don’t make sense to me; but it’s certainly true that the layer checks do take a big part of that.

Yes, copies are implemented in a way that assumes the destination needs to be updated; an extra “is the destination already exactly what it want to be” check would, for _most_ copies, be just an extra delay. OTOH for sync it’s probably the other way around, and it may well make sense to opt in to doing that check (maybe just assuming that the image does not need converting). Or, in the extreme, don’t ask the destination at all, and maintain a local state file that records the last manifest digest written to a particular tag. (We would probably still have to query each individual tag at the source, all 4000 or however many of them.)

This sync thing is starting to grow fairly complex…

**Comment by gjenkins8** [comment]
> So, going by this specific record, if I’m reading it right, reading the original manifest and config took 0 ms, ??? 
Ugh, yeah, I used the timestamps appended by our CI executor (Jenkins), which seems to not be super accurate. Apologies for not spotting that earlier.

**Comment by mtrmac** [comment]
That’s fine, the order-of-magnitude indication of how long it takes to “not copy” a single image is valuable on its own.

**Comment by gjenkins8** [comment]
> an extra “is the destination already exactly what we want it to be” check 

Would this be difficult to implement? If easy enough, and you can point me to roughly where the changes could be made. I could patch a local build and see what improvement this might make?

**Comment by eperdeme** [comment]
Reopened as on going discussion.  

A nice to have feature as mentioned above would be full sync validating all layers for all tags, basically what it does now and a fast sync which ignores layers if destination tag exists.  Could then do a fast hourly and full every night.  

Issues of course if tags are overwritten like latest and major version tags.

**Comment by mtrmac** [comment]
> Would this be difficult to implement? If easy enough, and you can point me to roughly where the changes could be made. I could patch a local build and see what improvement this might make?

That would be interesting. Very roughly, in `containers/image/copy.copyOneImage`, somewhere after `IsRunningImageAllowed` (exact placement TBD, probably affects correctness!, but this could be good enough for a basic performance estimate), compare that the byte arrays returned by `unparsedImage.Manifest()` and `c.dest.Reference().NewImageSource().GetManifest()` (+ appropriate error checking on all the calls),  and claim success if they are equal.

**Comment by gjenkins8** [comment]
nice, thanks @mtrmac ! Give me a few days to cycle this into my work and I'll hopefully be able to post results (or perhaps a question) shortly

**Comment by gjenkins8** [comment]
https://github.com/containers/image/pull/1041 reduces the sync of about 650 images from ~35mins to ~7mins (!) With roughly the simply change you described: fetch the destination manifest, and compare the bytes to the source manifest. Bail from the actual copy if the destination exists and compares equal.

So this seems like a nice performance win for the use case of synchronizing a large number of images which are expected to already exist.

The PR is still very much a draft, but this functionality would certainly be a good to have IMHO for this usage pattern.

![skopeo_sync](https://user-images.githubusercontent.com/2787248/92347089-b10ca780-f083-11ea-9c43-588e5a55d88a.png)

**Comment by mtrmac** [comment]
Thanks, that’s very encouraging. I’m not sure I’ll have the time to look into the details of the behavior this week, but it’s definitely an important improvement to prioritize.

**Comment by samdoran** [comment]
I can confirm with `skopeo` 1.2.3 sycing is _much_ faster because nothing is copied if it already exists in the destination. Our daily `skopeo sync` went from ~20 minutes to less than five minutes. 🎊 

<img width="673" alt="More Faster Sync" src="https://user-images.githubusercontent.com/4047861/117317488-90d1fa00-ae57-11eb-9966-5f0a2b4d194e.png">

**Comment by TomSweeneyRedHat** [comment]
@samdoran I'm curious, how do you run skopeo?  Do you install it from an rpm, or are you running it via one of the container images?

**Comment by TomSweeneyRedHat** [comment]
@samdoran and I forgot to add, TYVM for sharing the results!

**Comment by samdoran** [comment]
> Do you install it from an rpm, or are you running it via one of the container images?

It is [run from a specific image tag](https://github.com/ansible/bedrock/blob/d969f07ca8f64141df2b8a4cbbdf362b7b9b3734/azure-pipelines.yml#L35-L43) on Quay in Azure Pipelines. The runs are parallelized to reduce total run time.

**Comment by TomSweeneyRedHat** [comment]
Thanks @samdoran

**Comment by eperdeme** [comment]
Closing as graphs suggest good gains since the PR. Tested it our side and we do see the improved performance.

Great work!
